certbot-dns-digitalocean/certbot_dns_digitalocean/dns_digitalocean.py

```diff
@@ -0,0 +1,168 @@
+
+    def del_txt_record(self, domain_name, record_name, record_content):
+        """
+        Delete a TXT record using the supplied information.
+
+        Note that both the record's name and content are used to ensure that similar records
+        created concurrently (e.g., due to concurrent invocations of this plugin) are not deleted.
+
+        Failures are logged, but not raised.
+
+        :param str domain_name: The domain to use to associate the record with.
+        :param str record_name: The record name (typically beginning with '_acme-challenge.').
+        :param str record_content: The record content (typically the challenge validation).
+        """
+
+        try:
+            domain = self._find_domain(domain_name)
+        except digitalocean.Error as e:
+            logger.debug('Error finding domain using the DigitalOcean API: %s', e)
+            return
+
+        try:
+            domain_records = domain.get_records()
+
+            matching_records = [record for record in domain_records
+                                if record.type == 'TXT'
+                                and record.name == self._compute_record_name(domain, record_name)
+                                and record.data == record_content]
+        except digitalocean.Error as e:
+            logger.debug('Error getting DNS records using the DigitalOcean API: %s', e)
+            return
+
+        for record in matching_records:
+            try:
+                logger.debug('Removing TXT record with id: %s', record.id)
+                record.destroy()
+            except digitalocean.Error as e:
+                logger.warn('Error deleting TXT record %s using the DigitalOcean API: %s',
+                            record.id, e)
+
+    def _find_domain(self, domain_name):
+        """
+        Find the domain object for a given domain name.
+
+        :param str domain_name: The domain name for which to find the corresponding Domain.
+        :returns: The Domain, if found.
+        :rtype: `~digitalocean.Domain`
+        :raises certbot.errors.PluginError: if no matching Domain is found.
+        """
+
+        domain_name_guesses = dns_common.base_domain_name_guesses(domain_name)
+
+        domains = self.manager.get_all_domains()
+
+        for guess in domain_name_guesses:
+            matches = [domain for domain in domains if domain.name == guess]
+
+            if len(matches) > 0:
+                domain = matches[0]
+                logger.debug('Found base domain for %s using name %s', domain_name, guess)
+                return domain
+
+        raise errors.PluginError('Unable to determine base domain for {0} using names: {1}.'
+                                 .format(domain_name, domain_name_guesses))
+
+    @staticmethod
+    def _compute_record_name(domain, full_record_name):
+        # The domain, from DigitalOcean's point of view, is automatically appended.
+        return full_record_name.split("." + domain.name)[0]
```

Comment:
While silly, if the user attempts to validate a name containing the domain from Digital Ocean's point of view multiple times this doesn't work. For example, Digital Ocean sees my domain as `example.org`, but if I attempt to get a cert for `example.org.example.org` I fail the challenge.

I believe we can work around the problem by changing this to:
```
return full_record_name.rpartition("." + domain.name)[0]
```